### ISSUE TITLE
reportOverlappingOverload changes between 1.1.337 -> 1.1.338 in the presence of @staticmethod

### ISSUE THREAD

**Issue by insilications** [microsoft/pyright#6585]
Under Pyright 1.1.337 and latest Pylance preview, the following:
```python
from __future__ import annotations
from typing import overload, Callable, Any

class MyClass:

    @overload
    def connect(self, arg__1: bytes, arg__2: Callable[..., Any], /, type: str = ...) -> str: ...
    @overload
    @staticmethod
    def connect(arg__1: MyClass, arg__2: bytes, arg__3: Callable[..., Any], /, type: str = ...) -> str: ...
```

Gives the following error: `Overload 2 for "connect" will never be used because its parameters overlap overload 1 (reportOverlappingOverload)`

But on the latest Pyright 1.1.338, the error goes away. This kind of stub typing is usually present on PySide6, which uses extended types and a special C implemented `__getattribute__` to solve the obvious issue of mixing static methods and instance methods. I previously used the `reportOverlappingOverload` error to automatically append a `# pyright:
ignore[reportOverlappingOverload]` to offending declarations, but now that error is gone. Is that the correct behavior going further?

**Comment by erictraut** [comment]
I think it should be considered an error if overloads don't consistently use `@staticmethod`. Mypy issues an error in this case. Pyright should as well. It never occurred to me that someone would attempt to abuse overloads in this manner, so I didn't think to add a check for this. I've filed [this issue](https://github.com/microsoft/pyright/issues/6592) to track this.

You shouldn't rely on the absence or presence of any other errors when the overload definitions are incorrect like this.

**Comment by insilications** [comment]
> It never occurred to me that someone would attempt to abuse overloads in this manner

A number of relevant projects are inserting strategic extension modules that implement new types with custom special method behavior. Forgive my ignorance, as I've only recently started delving deeper into Python, but it seems to me that in a distant future the Python typing experience should be able to express and accommodate for that.


> You shouldn't rely on the absence or presence of any other errors when the overload definitions are incorrect like this.

I agree. I used this solution as a crutch to detect when overloads use @staticmethod or @classmethod inconsistently, get the Pyright json output and patch the stubs. Pyright is already my favorite and has very powerful inference. I didn't want to parse the project stub using libcst just to make sure I'm capturing all the edge cases.
